pkg/apis/picchu/v1alpha1/common.go

```diff
@@ -19,10 +20,64 @@ var (
 	PortLocal PortMode = "local"
 )
 
+const (
+	HumaneSchedule = "humane"
+
+	RateIncrementDefault   = int32(10)
+	RateDelayDefault       = time.Duration(6) * time.Second
+	ReleaseMaxDefault      = int32(100)
+	ReleaseScheduleDefault = HumaneSchedule
+)
+
 type PortInfo struct {
 	Name          string          `json:"name,omitempty"`
+	Hosts         []string        `json:"hosts,omitempty"`
 	Port          int32           `json:"port,omitempty"`
 	ContainerPort int32           `json:"containerPort,omitempty"`
 	Protocol      corev1.Protocol `json:"protocol,omitempty"`
 	Mode          PortMode        `json:"mode"`
 }
+
+type ScaleInfo struct {
+	Min       int32        `json:"min,omitempty"`
+	Default   int32        `json:"default,omitempty"`
+	Max       int32        `json:"max,omitempty"`
+	Resources ResourceInfo `json:"resources,omitempty"`
+}
+
+type ResourceInfo struct {
+	CPU string `json:"cpu,omitempty"`
+}
+
+type ReleaseInfo struct {
+	Eligible bool     `json:"eligible,omitempty"`
+	Max      int32    `json:"max,omitempty"`
+	Rate     RateInfo `json:"rate,omitempty"`
+	Schedule string   `json:"schedule,omitempty"`
+}
+
+func (r *ReleaseInfo) GetMax() int32 {
```

Comment:
we can use [defaulter-gen](https://github.com/kudzu-sh/api/blob/master/Makefile#L29-L30) for this. you declare some [`SetDefaults` functions](https://github.com/kudzu-sh/api/blob/master/kudzu/v1alpha1/source_defaults.go) for spec types and their descendants (like `ReleaseInfo`, `RateInfo`), and it generates [glue code](https://github.com/kudzu-sh/api/blob/master/kudzu/v1alpha1/zz_generated.defaults.go) for all the layers of your CRD types – I only defined a defaulter for `ImageSpec` and it generated a defaulter for its parent `API` resource.

inside the reconciler, you can then call [`scheme.Default`](https://godoc.org/k8s.io/apimachinery/pkg/runtime#Scheme.Default) on a resource, it’ll populate all defaults, and we won’t need `Get` wrapper methods.